# Python Programming: Intermediate

<img src="../images/OOP_1.jpg" alt="Python" style="width: 400px;"/>

# Program so far..
***
- Python Basics
- Python Programming Constructs
- Data Structures


# What are we going to learn today?
***
- Modular Programming in Python
    - Functions
    - Brief on regular exceptions
    - Object Oriented Programming
- NumPy

# Modular Programming in Python
***
- Like all programming languages, Python provides various constructs to enable code reuse <br/><br/>

- Python provides functions/methods, classes and modules for reuse

<img src="../images/icon/Technical-Stuff.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />

## What is a function?
***

Formally, a function is a useful device that groups together a set of statements so they can be run more than once. They can also let us specify parameters that can serve as inputs to the functions.

On a more fundamental level, functions allow us to not have to repeatedly write the same code again and again. If you remember back to the lessons on strings and lists, remember that we used a function len() to get the length of a string. Since checking the length of a sequence is a common task you would want to write a function that can do this repeatedly at command.

Functions will be one of most basic levels of reusing code in Python, and it will also allow us to start thinking of program design (we will dive much deeper into the ideas of design when we learn about Object Oriented Programming).


## Function Definition
***
In the syntax below:
- **def** is the keyword used to define functions
- arg1...argn, \*args and \*\*kwargs are function parameters (and are optional)
- The expression after the `return` keyword is the value returned to the caller (optional)

Now lets look at how we can create a function in python

In [2]:
def name_of_function(arg1,arg2):
    '''
    This is where the function's Document String (doc-string) goes
    '''
    # Do stuff here
    #return desired result

In [ ]:
sorted

In [1]:
def my_func(param1='default'):
    """
    Docstring goes here.
    """
    print(param1)

In [2]:
my_func('Python')

Python


In [ ]:
my_func

In [4]:
def my_func(param1):
    """
    Docstring goes here.
    """
    print(param1)

In [5]:
my_func()

TypeError: my_func() missing 1 required positional argument: 'param1'

## def Statements

We begin with def then a space followed by the name of the function. Try to keep names relevant, for example len() is a good name for a length() function. Also be careful with names, you wouldn't want to call a function the same name as a built-in function in Python (such as len).

Next come a pair of parenthesis with a number of arguments separated by a comma. These arguments are the inputs for your function. You'll be able to use these inputs in your function and reference them. After this you put a colon.

Now here is the important step, you must indent to begin the code inside your function correctly. Python makes use of whitespace to organize code. Lots of other programing languages do not do this, so keep that in mind.

Next you'll see the doc-string, this is where you write a basic description of the function. Using iPython and iPython Notebooks, you'll be ab;e to read these doc-strings by pressing Shift+Tab after a function name. Doc strings are not necessary for simple functions, but its good practice to put them in so you or other people can easily understand the code you write.

After all this you begin writing the code you wish to execute.

The best way to learn functions is by going through examples. So let's try to go through examples that relate back to the various objects and data structures we learned about before.

### Example 1: A simple print 'hello' function
In this example we take our first steps to write a working function which prints out 'hello'

In [2]:
def say_hello():
    print 'hello'

#call the function 

say_hello()

hello


## Function Arguments
***
- Parameters may or may not have default values (in the example, argn has default value 3)
- A function parameter can be passed either by position or by key/name
- A function can accept a variable number of positional arguments (\*args)
- A function can accept a variable number of keyword arguments (\*\*kwargs)
- Function parameters are passed by object reference


In [3]:
def function_name(arg1, arg2, argn=3, *args, **kwargs):
    print("This is a function.")
    # Function body here
    return value

### Default Arguments

Default arguements are those arguments which can be changed but their value remains the same if not given. An example of the same is given below.

In [4]:
def double_the_number(num=1):
    return num * 2

print(double_the_number())

2


### Default Argument should start from right to left
***

In [12]:
def multiply(num1, num2, num3=10):
    print(num1*num2*num3)

In [13]:
multiply(10, 30, 40)

12000


<img src="../images/icon/ppt-icons.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />

### Mini Challenge
***
Its your turn now. Can you write a function square_root having default value of input a=9 and find the square-root of the same.

### Variable Arguments (1/2)

The special syntax *args in function definitions in python is used to pass a variable number of arguments to a function. It is used to pass a non-keyworded, variable-length argument list.

- The syntax is to use the symbol * to take in a variable number of arguments; by convention, it is often used with the word args.
- What *args allows you to do is take in more arguments than the number of formal arguments that you previously defined. With *args, any number of extra arguments can be tacked on to your current formal parameters (including zero extra arguments).

- For example : we want to make a multiply function that takes any number of arguments and able to multiply them all together. It can be done using *args.
Using the *, the variable that we associate with the * becomes an iterable meaning you can do things like iterate over it, run some higher order functions such as map and filter, etc. The example of the same is shown below.

In [18]:
def print_positional_arguments(num1, *args):
    
    print(type(args))
    print(args)
    print(len(args))

In [17]:
print_positional_arguments(1, 2, 3, 4, 'Hello')

<class 'tuple'>
(2, 3, 4, 'Hello')
4


### Variable Arguments (2/2)
The special syntax `**kwargs` in function definitions in python is used to pass a keyworded, variable-length argument list. We use the name kwargs with the double star. The reason is because the double star allows us to pass through keyword arguments (and any number of them).

- A keyword argument is where you provide a name to the variable as you pass it into the function.
- We can think of the kwargs as being a dictionary that maps each keyword to the value that we pass alongside it. That is why when we iterate over the kwargs there doesn’t seem to be any order in which they were printed out.

An Example of the same is shown below.


In [29]:
def print_keyword_arguments(num1, **kwargs):
    
    print(type(kwargs))    
    print(kwargs)
    print(type(kwargs.keys()))
    print(type(kwargs.values()))
    print(' ')
    print(kwargs.items())

    #return kwargs.values()


In [30]:
print_keyword_arguments(1, num2=2, str3='String Input')

<class 'dict'>
{'num2': 2, 'str3': 'String Input'}
<class 'dict_keys'>
<class 'dict_values'>
 
dict_items([('num2', 2), ('str3', 'String Input')])


## Return Values
***

A Return value is a value that is returned after performing a specific operation in a function. Some advantages and usecases of return values are as follows.
- Unlike some other languages, Python allows returning multiple values 
- However, the multiple values is just a tuple of values
- Because of this, the tuple can simply be 'opened' into multiple variables


Now lets look at a function returning multiple values.

In [32]:
def square_and_cube(num):
    return num**2, num**3

In [34]:
answer = square_and_cube(3)
print(answer)
print(type(answer))

(9, 27)
<class 'tuple'>


In [35]:
square, cube = square_and_cube(3)
print(square, cube)

9 27


<img src="../images/icon/ppt-icons.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />
### Mini Challenge
***
Its your turn again. Can you write a function function_tuple that takes in two numbers a and b and returns the sum as well as the product of these numbers

### Introduce Random Number concept

In [36]:
import random

In [38]:
#dir(random)

In [40]:
random.random()

0.510946239334319

In [43]:
print(random.randint(1, 100))

34


1. Factorial Number
2. Guess a Number
3. Print if a number is prime number

<img src="../images/icon/Concept-Alert.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />

## Lambda
***
One of Pythons most useful (and for beginners, confusing) tools is the lambda expression. lambda expressions allow us to create "anonymous" functions. This basically means we can quickly make ad-hoc functions without needing to properly define a function using def.

Function objects returned by running lambda expressions work exactly the same as those created and assigned by defs. There is key difference that makes lambda useful in specialized roles:

**Lambda's body is a single expression, not a block of statements.**
* Python supports the creation of anonymous functions at runtime, using a construct called **`lambda`**
* This approach is most commonly used when passing a simple function as an argument to another function.
* Lambdas are generally used in conjunction with typical functional concepts like `filter()`, `map()` and `reduce()`.

Now Lets slowly break down a lambda expression by deconstructing a function:

In [1]:
def square(num):
    result = num**2
    return result
square(2)

4

We can actually write this in one line (although it would be bad style to do so)

In [2]:
def square(num): return num**2

This is the form a function that a lambda expression intends to replicate. A lambda expression can then be written as:

In [3]:
square_lambda = lambda num: num**2

square_lambda(2)

4

<img src="../images/icon/ppt-icons.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />
### Mini Challenge
***
Can you write a lambda function to add two numbers a and b. 

<img src="../images/icon/Technical-Stuff.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />
## Higher order functions
***
In Python, we treat functions as first class objects, allowing you to perform the operations on functions.

In [1]:
def add_lambda(num1, num2):
    return num1*num2

In [2]:
# this is a higher order function

def calculate(func, num1, num2):
    return func(num1, num2)

# call calculate with the add function
calculate(add_lambda, 1, 2)

2

<img src="../images/icon/Technical-Stuff.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />
### `map`
***
map() is a function that takes in two arguments: a function and a sequence iterable. In the form: map(function, sequence) **`map(function, sequence)`**

The first argument is the name of a function and the second a sequence (e.g. a list). map() applies the function to all the elements of the sequence. It returns a new list with the elements changed by function.

When we went over list comprehension we created a small expression to convert Fahrenheit to Celsius. Let's do the same here but use map.

We'll start with two functions:

In [ ]:
seq = [2, 4, 6, 8]

def times2(param):
    return param*2

print (map(times2, seq))

print (list(map(times2,seq)))

print(map(lambda var: var*2,seq))

In [6]:
def fahrenheit(T):
    return ((float(9)/5)*T + 32)

def celsius(T):
    return (float(5)/9)*(T-32)
    
temp = [0, 22.5, 40,100]

Now lets see map() in action:

In [7]:
F_temps = map(fahrenheit, temp)

#Show
F_temps

[32.0, 72.5, 104.0, 212.0]

In [8]:
# Convert back
map(celsius, F_temps)

[0.0, 22.5, 40.0, 100.0]

In the example above we haven't used a lambda expression. By using lambda, we wouldn't have had to define and name the functions fahrenheit() and celsius().

In [9]:
map(lambda x: (5.0/9)*(x - 32), F_temps)

[0.0, 22.5, 40.0, 100.0]

<img src="../images/icon/ppt-icons.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />
### Mini Challenge
***
Great! We got the same result! Using map is much more commonly used with lambda expressions since the entire purpose of map() is to save effort on having to create manual for loops.

map() can be applied to more than one iterable. The iterables have to have the same length.

For example lets map a lambda expression:


In [31]:
a = [1,2,3,4]
b = [5,6,7,8]
c = [9,10,11,12]

In [32]:
#list(map(lambda x, y, z: x+y+z, a, b, c))

[15, 18, 21, 24]

<img src="../images/icon/Technical-Stuff.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />
### `filter`
***

The function filter(function, list) offers a convenient way to filter out all the elements of an iterable, for which the function returns True. 

The function filter(function(),l) needs a function as its first argument. The function needs to return a Boolean value (either True or False). This function will be applied to every element of the iterable. Only if the function returns True will the element of the iterable be included in the result.

Lets see some examples:


In [30]:
seq = [1,2,3,4,5]
list(filter(lambda item: item%2 == 0, seq))

[2, 4]

In [11]:
numbers = list(range(10))

def is_even(x):
    return x % 2 == 0

is_even_lambda = lambda x: x % 2 == 0

print(filter(is_even, numbers))
print(filter(is_even_lambda, numbers))

[0, 2, 4, 6, 8]
[0, 2, 4, 6, 8]


### Global and non local variable
***

In [65]:
globvar = 10
def read1():
    print(globvar)
def write1():
    global globvar
    globvar = 5
def write2():
    globvar = 15

read1()
write1()
read1()
write2()
read1()

10
5
5


In [81]:
def outer_function():
    a = 5
    def inner_function():
        print(a) # you can print/access outer variable but
        #a += 5 # you can bot modify it
        #a = 10 # to modify a outer variable, you need to refer it with non-local keyword
        print("Inner function: ", a)
    inner_function()
    print("Outer function: ", a)

outer_function()

5
Inner function:  5
Outer function:  5


In [70]:
def outer_function():
    a = 5
    print('outer function: ', a)
    def inner_function():
        nonlocal a # refering to a defined in outer_function
        a = 10
        print("Inner function: ", a)
    inner_function()
    print("Outer function: ", a)

outer_function()

outer function:  5
Inner function:  10
Outer function:  10


<img src="../images/icon/Concept-Alert.png" alt="Concept-Alert" style="width: 100px;float:left; margin-right:15px"/>
<br />
## Nested Functions and Closures
***
- Python functions can be defined within the scope of another function.

- The inner function **definition** happens only when the outer function executes

- The inner function is only in scope inside the outer function, so it is often most useful when the inner function is being returned (or passed to another function)

- It is possible to return an inner function that "remembers" the state of the outer function has completed execution. This is called a closure.

In [84]:
# Nested Functions - function returning function

def print_msg(msg):
    # This is the outer enclosing function

    def printer():
    # This is the nested function
        print(msg) # Very similar to non-local variable

    printer()

# We execute the function
# Output: Hello
print_msg("Hello")

# We can see that nested function printer() was able to access the non-local variable msg of the enclosing function

Hello


- In the example above, what would happen if the last line of the function **print_msg()** returned the printer() - function instead of calling it? 
- This means the function was defined as follows.

In [85]:
def print_msg(msg):
    # This is the outer enclosing function

    def printer():
    # This is the nested function
        print(msg)

    return printer  # this got changed

# Now let's try calling this function.
# Output: Hello
another = print_msg("Hello")
another()

Hello


- ** On calling another(), the message was still remembered although we had already finished executing the print_msg() function **

- This technique by which some data ("Hello") gets attached to the code is called ** closure in Python.**
- This value in the enclosing scope is remembered even when the **variable goes out of scope** or the **function itself is removed from the current namespace**.

In [86]:
del print_msg

In [87]:
another()

Hello


In [88]:
print_msg()

NameError: name 'print_msg' is not defined

In [12]:
# Nested Functions - function returning function
def outer(a):
    def inner(b):
        return a + b + 5
    return inner

twentyfive_adder = outer(20)
seven_adder = outer(2)

print(twentyfive_adder(5))
print(seven_adder(5))

30
12


** So what are closures good for? **

- Closures can avoid the use of global values and provides some form of data hiding. 
- It can also provide an object oriented solution to the problem.
- When there are few methods (one method in most cases) to be implemented in a class, closures can provide an alternate and more elegant solutions. But when the number of attributes and methods get larger, better implement a class.

** Use lambda expressions and the filter() function to filter out words from a list that don't start with the letter 's'. For example: **

    seq = ['soup','dog','salad','cat','great']

**should be filtered down to:**

    ['soup','salad']
***

** file and dictionary and functions - read how many time a name appears in a file **
http://www.practicepython.org/assets/nameslist.txt
***


In [34]:
file_name = 'nameslist.txt'
f = open(file_name, 'r')
data = f.read()

In [36]:
all_names = data.split('\n')
print(all_names)

['Darth', 'Luke', 'Darth', 'Lea', 'Darth', 'Lea', 'Lea', 'Luke', 'Darth', 'Lea', 'Darth', 'Darth', 'Lea', 'Lea', 'Darth', 'Lea', 'Darth', 'Lea', 'Luke', 'Darth', 'Lea', 'Lea', 'Darth', 'Lea', 'Darth', 'Darth', 'Lea', 'Lea', 'Luke', 'Luke', 'Lea', 'Darth', 'Darth', 'Luke', 'Lea', 'Darth', 'Darth', 'Lea', 'Lea', 'Lea', 'Lea', 'Lea', 'Luke', 'Darth', 'Luke', 'Lea', 'Lea', 'Lea', 'Lea', 'Luke', 'Lea', 'Darth', 'Lea', 'Lea', 'Darth', 'Lea', 'Lea', 'Darth', 'Darth', 'Lea', 'Darth', 'Lea', 'Darth', 'Luke', 'Lea', 'Luke', 'Darth', 'Darth', 'Luke', 'Darth', 'Lea', 'Darth', 'Lea', 'Luke', 'Lea', 'Lea', 'Lea', 'Lea', 'Lea', 'Darth', 'Lea', 'Lea', 'Lea', 'Lea', 'Lea', 'Lea', 'Lea', 'Luke', 'Lea', 'Lea', 'Lea', 'Lea', 'Lea', 'Lea', 'Darth', 'Luke', 'Darth', 'Lea', 'Lea', 'Darth']


In [38]:
def return_name_count(names):
    print(type(names))
    name_dict = {}
    for name in names:
        if name in name_dict.keys():
            name_dict[name] += 1
        else:
            name_dict[name] = 1
            
    return name_dict

In [39]:
name_dict = return_name_count(all_names)

<class 'list'>


In [40]:
name_dict

{'Darth': 31, 'Lea': 54, 'Luke': 15}

In [47]:
# Default sort by keys
sorted(name_dict.items())

[('Darth', 31), ('Lea', 54), ('Luke', 15)]

In [48]:
sorted(name_dict.items(), reverse=True)

[('Luke', 15), ('Lea', 54), ('Darth', 31)]

In [56]:
sorted(name_dict, key = name_dict.get)

['Luke', 'Darth', 'Lea']

In [63]:
s_list = [(k, name_dict[k]) for k in sorted(name_dict, key = name_dict.get)]
s_list

[('Luke', 15), ('Darth', 31), ('Lea', 54)]

In [64]:
s_dict = {k: name_dict[k] for k in sorted(name_dict, key = name_dict.get)}
s_dict

{'Darth': 31, 'Lea': 54, 'Luke': 15}

# Regular Expressions
***

- Regular expressions are a powerful language for matching text patterns
- The Python "re" module provides regular expression support

In Python a regular expression search is typically written as
- ** match = re.search(pat, str)**

- The re.search() method takes a regular expression pattern and a string and searches for that pattern within the string. 
- If the search is successful, search() returns a match object or None otherwise. 
- Therefore, the search is usually immediately followed by an if-statement to test if the search succeeded.
- Example pattern 'word:' followed by a 3 letter word (details below):

In [91]:
import re

In [92]:
str = 'an example word:cat!!'
match = re.search(r'word:\w\w\w', str)

# If-statement after search() tests if it succeeded
if match:                      
    print('found', match.group()) ## 'found word:cat'
else:
    print('did not found')

found word:cat


### Basic Patterns
***

The power of regular expressions is that they can specify patterns, not just fixed characters. Here are the most basic patterns which match single chars:

- **a, X, 9**, < -- ordinary characters just match themselves exactly. The meta-characters which do not match themselves because they have special meanings are: . ^ $ * + ? { [ ] \ | ( ) (details below)
- **.** (a period) -- matches any single character except newline '\n'
- **\w** -- (lowercase w) matches a "word" character: a letter or digit or underbar [a-zA-Z0-9_]. Note that although "word" is the mnemonic for this, it only matches a single word char, not a whole word. \W (upper case W) matches any non-word character.
- **\b** -- boundary between word and non-word
- **\s** -- (lowercase s) matches a single whitespace character -- space, newline, return, tab, form [ \n\r\t\f]. \S (upper case S) matches any non-whitespace character.
- **\t, \n, \r** -- tab, newline, return
- **\d** -- decimal digit [0-9] (some older regex utilities do not support but \d, but they all support \w and \s)
- **^ = start, `$` = end** -- match the start or end of the string
- **\** -- inhibit the "specialness" of a character. So, for example, use \. to match a period or \\ to match a slash. If you are unsure if a character has special meaning, such as '@', you can put a slash in front of it, \@, to make sure it is treated just as a character.

### The basic rules of regular expression search for a pattern within a string are:

- The search proceeds through the string from start to end, stopping at the first match found
- All of the pattern must be matched, but not all of the string
- If match = re.search(pat, str) is successful, match is not None and in particular match.group() is the matching text

In [97]:
import re

## Search for pattern 'iii' in string 'piiig'.
## All of the pattern must match, but it may appear anywhere.
## On success, match.group() is matched text.
match = re.search(r'iii', 'piiig') #=>  found, match.group() == "iii"
print(match.group())

match = re.search(r'igs', 'piiig') #=>  not found, match == None
print(match)

## . = any char but \n
match = re.search(r'..g', 'piiig') #=>  found, match.group() == "iig"
print(match.group())

## \d = digit char, \w = word char
match = re.search(r'\d\d\d', 'p123g') #=>  found, match.group() == "123"
print(match.group())

match = re.search(r'\w\w\w', '@@abcd!!') #=>  found, match.group() == "abc"
print(match.group())

iii
None
iig
123
abc


### Repetition
#### Things get more interesting when you use + and * to specify repetition in the pattern

- **+** -- 1 or more occurrences of the pattern to its left, e.g. 'i+' = one or more i's
- **\* ** -- 0 or more occurrences of the pattern to its left
- **?** -- match 0 or 1 occurrences of the pattern to its left

In [101]:
## i+ = one or more i's, as many as possible.
match = re.search('pi+', 'piiig') #=>  found, match.group() == "piii"
print(match.group())

## Finds the first/leftmost solution, and within it drives the +
## as far as possible (aka 'leftmost and largest').
## In this example, note that it does not get to the second set of i's.
match = re.search('i+', 'piigiiii') #=>  found, match.group() == "ii"
print(match.group())

## \s* = zero or more whitespace chars
## Here look for 3 digits, possibly separated by whitespace.
match = re.search('\d\s*\d\s*\d', 'xx1 2   3xx') #=>  found, match.group() == "1 2   3"
print(match.group())

match = re.search(r'\d\s*\d\s*\d', 'xx12  3xx') #=>  found, match.group() == "12  3"
print(match.group())

match = re.search(r'\d\s*\d\s*\d', 'xx123xx') #=>  found, match.group() == "123"
print(match.group())

## ^ = matches the start of string, so this fails:
match = re.search(r'^b\w+', 'foobar') #=>  not found, match == None
#print match.group()

## but without the ^ it succeeds:
match = re.search(r'b\w+', 'foobar') #=>  found, match.group() == "bar"
print(match.group())

piii
ii
1 2   3
12  3
123
bar


### Emails Example

In [103]:
str = 'purple rk-v@gmail.com monkey dishwasher'
match = re.search(r'\w+@\w+', str)
if match:
    print(match.group())  ## 'b@google'

v@gmail


#### Square Brackets

- Square brackets can be used to indicate a set of chars, so [abc] matches 'a' or 'b' or 'c'
- The codes \w, \s etc. work inside square brackets too with the one exception that dot (.) just means a literal dot
- For the emails problem, the square brackets are an easy way to add '.' and '-' to the set of chars which can appear around the @ with the pattern **r'[\w.-]+@[\w.-]+'** to get the whole email address:

In [17]:
match = re.search(r'[\w.-]+@[\w.-]+', str)
if match:
    print(match.group())  ## 'alice-b@google.com'

rk-v@gmail.com


- You can also use a dash to indicate a range, so [a-z] matches all lowercase letters. 
- To use a dash without indicating a range, put the dash last, e.g. [abc-]. 
- An up-hat (^) at the start of a square-bracket set inverts it, so [^ab] means any char except 'a' or 'b'.

### Group Extraction

- The "group" feature of a regular expression allows you to pick out parts of the matching text. 
- Suppose for the emails problem that we want to extract the **username** and **host** separately. 
- To do this, add parenthesis ( ) around the username and host in the pattern, like this: **r'([\w.-]+)@([\w.-]+)'.**
- In this case, the parenthesis do not change what the pattern will match, instead they establish logical "groups" inside of the match text.
- On a successful search, match.group(1) is the match text corresponding to the 1st parenthesis, and match.group(2) is the text corresponding to the 2nd parenthesis. 
- The plain match.group() is still the whole match text as usual.

In [107]:
str = 'purple rk-v@gmail.com monkey dishwasher'
match = re.search('([\w.-]+)@([\w.-]+)', str)
if match:
    print(match.group())   ## 'alice-b@google.com' (the whole match)
    print(match.group(1)) ## 'alice-b' (the username, group 1)
    print(match.group(2))  ## 'google.com' (the host, group 2)

rk-v@gmail.com
rk-v
gmail.com


In [118]:
match.start(), match.end()

(7, 21)

### findall

- findall() is probably the single most powerful function in the re module. Above we used re.search() to find the first match for a pattern.
- findall() finds *all* the matches and returns them as a list of strings, with each string representing one match

In [109]:
## Suppose we have a text with many email addresses
str = 'purple alice@google.com, blah monkey bob@abc.com blah dishwasher'

## Here re.findall() returns a list of all the found email strings
emails = re.findall(r'[\w\.-]+@[\w\.-]+', str) ## ['alice@google.com', 'bob@abc.com']
for email in emails:
    # do something with each found email string
    print(email)

alice@google.com
bob@abc.com


### findall With Files

- For files, you may be in the habit of writing a loop to iterate over the lines of the file, and you could then call findall() on each line.
- Instead, let findall() do the iteration for you -- much better! Just feed the whole file text into findall() and let it return a list of all the matches in a single step (recall that f.read() returns the whole text of a file in a single string):

In [110]:
# Open file
f = open('/home/rajneeshkumar/Downloads/google-python-exercises/NOTICE.txt', 'r')
# Feed the file text into findall(); it returns a list of all the found strings
strings = re.findall(r'some pattern', f.read())

FileNotFoundError: [Errno 2] No such file or directory: '/home/rajneeshkumar/Downloads/google-python-exercises/NOTICE.txt'

### findall and Groups

- The parenthesis ( ) group mechanism can be combined with findall(). 
- If the pattern includes 2 or more parenthesis groups, then instead of returning a list of strings, findall() returns a list of *tuples*. 
- Each tuple represents one match of the pattern, and inside the tuple is the group(1), group(2) .. data. 
- So if 2 parenthesis groups are added to the email pattern, then findall() returns a list of tuples, each length 2 containing the username and host, e.g. ('alice', 'google.com').

In [116]:
str1 = 'purple alice@google.com, blah monkey bob@abc.com blah dishwasher'
tuples = re.findall(r'([\w\.-]+)@([\w\.-]+)', str1)
print(tuples)  ## [('alice', 'google.com'), ('bob', 'abc.com')]
for tuple in tuples:
    print(tuple[0]),  ## username
    print(tuple[1])  ## host

[('alice', 'google.com'), ('bob', 'abc.com')]
alice
google.com
bob
abc.com


In [120]:
### findall without groups

str = 'purple alice@google.com, blah monkey bob@abc.com blah dishwasher'
tuples = re.findall(r'[\w\.-]+@[\w\.-]+', str)
print(tuples)  ## [('alice', 'google.com'), ('bob', 'abc.com')]

['alice@google.com', 'bob@abc.com']


In [121]:
### findall withone group

str = 'purple alice@google.com, blah monkey bob@abc.com blah dishwasher'
tuples = re.findall(r'([\w\.-]+@[\w\.-]+)', str)
print(tuples)  ## [('alice', 'google.com'), ('bob', 'abc.com')]

['alice@google.com', 'bob@abc.com']


### Substitution (Advanced)

- The **re.sub(pat, replacement, str)** function searches for all the instances of pattern in the given string, and replaces them. 
- The replacement string can include '\1', '\2' which refer to the text from group(1), group(2), and so on from the original matching text.
- Here's an example which searches for all the email addresses, and changes them to keep the user (\1) but have yahoo0.com as the host.

In [117]:
str = 'purple alice@google.com, blah monkey bob@abc.com blah dishwasher'
## re.sub(pat, replacement, str) -- returns new string with all replacements,
## \1 is group(1), \2 group(2) in the replacement
print(re.sub(r'([\w\.-]+)@([\w\.-]+)', r'\1@yahooo.com', str))
print(re.sub(r'([\w\.-]+)@([\w\.-]+)', r'hey-there@\2.com', str))

purple alice@yahooo.com, blah monkey bob@yahooo.com blah dishwasher
purple hey-there@google.com.com, blah monkey hey-there@abc.com.com blah dishwasher


#### There is re.match(pat, str) too

In [122]:
str = 'purple rk-v@gmail.com monkey dishwasher'
search = re.search('[\w.-]+@[\w.-]+', str)
if search:
    print(search.group())   ## 'alice-b@google.com' (the whole match)
    
match = re.match('[\w.-]+@[\w.-]+', str)
if match:
    print(match.group())   ## 'alice-b@google.com' (the whole match)
else:
    print('No Match')

rk-v@gmail.com
No Match


### Excercises

- 1. Write a Python program to check that a string contains only a certain set of characters (in this case a-z, A-Z and 0-9)
- 2. Write a Python program that matches a string that has an a followed by zero or more b's. 
- 3. Write a Python program that matches a word containing 'z', not start or end of the word
- 4. Write a Python program to match a string that contains only upper and lowercase letters, numbers, and underscores
- 5. Write a Python program to find all three, four, five characters long words in a string.

In [105]:
# ** 1 **
import re
def is_allowed_specific_char(text):
    patterns = r'[a-zA-Z0-9]*'
    match = re.search(patterns,  text)
    if match:
        return match.group()
    else:
        return 'Not matched!'

print(is_allowed_specific_char("ABCDEFabcdef123450")) 
print(is_allowed_specific_char("*&%@#!}{"))

ABCDEFabcdef123450



In [88]:
# ** 2 **
import re
def text_match(text):
    patterns = 'ab*?'
    match = re.search(patterns,  text)
    if match:
        return match.group()
    else:
        return 'Not matched!'

print(text_match("ac"))
print(text_match("abc"))
print(text_match("abbc"))


a
a
a


In [92]:
# ** 3 **
import re
def text_match(text):
    patterns = r"\Bz\B"
    match = re.search(patterns,  text)
    if match:
        return match.group()
    else:
        return('Not matched!')

print(text_match("The quick brown fox jumps over the lazy dog."))
print(text_match("Python Exercises."))

z
Not matched!


In [56]:
import re
def text_match(text):
    patterns = r"\bz\b"
    match = re.search(patterns,  text)
    if match:
        return match.group()
    else:
        return('Not matched!')

print(text_match("The quick brown fox jumps over the lazy dog."))
print(text_match("Python Exercises."))

Not matched!
Not matched!


In [76]:
# ** 4 **
import re
def text_match(text):
    patterns = '^[a-zA-Z0-9_]*$'
    match = re.search(patterns,  text)
    if match:
        return match.group()
    else:
        return('Not matched!')

print(text_match("The quick brown fox jumps over the lazy dog"))
print(text_match("Thequick brown"))
print(text_match("Python_Exercises_1"))

Not matched!
Not matched!
Python_Exercises_1


In [103]:
# ** 5 **
import re
text = 'The quick brown fox jumps over the lazy dog. google'
print(re.findall(r"\b\w{3,6}\b", text))

['The', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog', 'google']


- For more on patterns - https://www.tutorialspoint.com/python/python_reg_expressions.htm

<img src="../images/icon/Concept-Alert.png" alt="Concept-Alert" style="width: 100px;float:left; margin-right:15px"/>
<br />
# Object Oriented Programming
***


Object Oriented Programming (OOP) tends to be one of the major obstacles for beginners when they are first starting to learn Python.

There are many,many tutorials and lessons covering OOP so feel free to Google search other lessons, and I have also put some links to other useful tutorials online at the bottom of this Notebook.

For this lesson we will construct our knowledge of OOP in Python by building on the following topics:

* Objects
* Using the *class* keyword
* Creating class attributes
* Creating methods in a class
* Learning about Inheritance
* Learning about Special Methods for classes

Lets start the lesson by remembering about the Basic Python Objects. For example:

In [108]:
l = [1,2,3]

Remember how we could call methods on a list?

In [109]:
l.count(2)

1

What we will basically be doing in this lecture is exploring how we could create an Object type like a list. We've already learned about how to create functions. So lets explore Objects in general:

## Objects

In Python, everything is an object. Remember from previous lectures we can use type() to check the type of object something is:

In [15]:
print type(1)
print type([])
print type(())
print type({})

<type 'int'>
<type 'list'>
<type 'tuple'>
<type 'dict'>


In [123]:
isinstance(int, object)

True

## Object Oriented Programming

- Python is a multi-paradigm programming language. Meaning, it supports different programming approach.

- One of the popular approach to solve a programming problem is by creating objects. This is known as Object-Oriented Programming (OOP).

- Everything in python is an object. There are two main characteristics of an object:
    - attributes
    - behaviour/methods
    
- For example, Parrot is an object,

    - name, age, color are attributes
    - singing, dancing are behavior
    
- Same is true for everythin like human, employee, vehicle etc

### OOP Terminology

- **Class** − A user-defined prototype for an object that defines a set of attributes that characterize any object of the class. The attributes are data members (class variables and instance variables) and methods, accessed via dot notation.

- **Class variable** − A variable that is shared by all instances of a class. Class variables are defined within a class but outside any of the class's methods. Class variables are not used as frequently as instance variables are.

- **Instance variable** − A variable that is defined inside a method and belongs only to the current instance of a class.

- **Inheritance** − The transfer of the characteristics of a class to other classes that are derived from it.

- **Instance** − An individual object of a certain class. An object obj that belongs to a class Circle, for example, is an instance of the class Circle.

- **Instantiation** − The creation of an instance of a class.

- **Method** − A special kind of function that is defined in a class definition.

- **Object** − A unique instance of a data structure that's defined by its class. An object comprises both data members (class variables and instance variables) and methods.

### OOPs Properties
- **Encapsulate** means to hide. Encapsulation is also called data hiding.You can think Encapsulation like a capsule (medicine tablet) which hides medicine inside it. Encapsulation is wrapping, just hiding properties and methods. Encapsulation is used for hide the code and data in a single unit to protect the data from the outside the world. Class is the best example of encapsulation.

- **Abstraction** refers to showing only the necessary details to the intended user. As the name suggests, abstraction is the "abstract form of anything". We use abstraction in programming languages to make abstract class. Abstract class represents abstract view of methods and properties of class.

- ** Polymorphism ** A concept of using common operation in different ways for different data input

- **Inheritance** Reusablity, inheriting properties from more generic to specific classes

So we know all these things are objects, so how can we create our own Object types? That is where the class keyword comes in.

## class
The user defined objects are created using the class keyword. The class is a blueprint that defines a nature of a future object. From classes we can construct instances. An instance is a specific object created from a particular class. For example, above we created the object 'l' which was an instance of a list object.


### Object Oriented Programming has great advantages over other programming styles:

- **Code Reuse and Recycling:** Objects created for Object Oriented Programs can easily be reused in other programs.
- **Encapsulation:** Once an Object is created, knowledge of its implementation is not necessary for its use. In older programs, coders needed understand the details of a piece of code before using it (in this or another program).
- **Abstraction:** Objects have the ability to hide certain parts of themselves from programmers. This prevents programmers from tampering with values they shouldn't. Additionally, the object controls how one interacts with it, preventing other kinds of errors. For example, a programmer (or another program) cannot set the width of a window to -400.
- **Design Benefits:** Large programs are very difficult to write. Object Oriented Programs force designers to go through an extensive planning phase, which makes for better designs with less flaws. In addition, once a program reaches a certain size, Object Oriented Programs are actually easier to program than non-Object Oriented ones.
- **Software Maintenance:** Programs are not disposable. Legacy code must be dealt with on a daily basis, either to be improved upon (for a new version of an exist piece of software) or made to work with newer computers and software. An Object Oriented Program is much easier to modify and maintain than a non-Object Oriented Program. So although a lot of work is spent before the program is written, less work is needed to maintain it over time.

**Let see how we can use class:**
***

We can think of class as an sketch of a parrot with labels. It contains all the details about the name, colors, size etc. Based on these descriptions, we can study about the parrot. Here, parrot is an object.

The example for class of parrot can be :


In [124]:
class Parrot:
    pass

In [126]:
# Create a new object type called Sample
class Sample(object):
    pass

# Instance of Sample
x = Sample()

print(type(x))

<class '__main__.Sample'>


By convention we give classes a name that starts with a capital letter. Note how x is now the reference to our new instance of a Sample class. In other words, we **instantiate** the Sample class.

Inside of the class we currently just have pass. But we can define class attributes and methods.

An **attribute** is a characteristic of an object.
A **method** is an operation we can perform with the object.

For example we can create a class called Dog. An attribute of a dog may be its breed or its name, while a method of a dog may be defined by a .bark() method which returns a sound.

Let's get a better understanding of attributes through an example.

## Attributes
The syntax for creating an attribute is:
    
    self.attribute = something
    
There is a special method called:

    __init__()

This method is used to initialize the attributes of an object. For example:

In [127]:
class Dog(object):
    def __init__(self, breed):
        self.breed = breed
        
sam = Dog(breed='Lab')
frank = Dog(breed='Huskie')

Lets break down what we have above.The special method

    __init__() 

is called automatically right after the object has been created:

    def __init__(self, breed):
Each attribute in a class definition begins with a reference to the instance object. It is by convention named self. The breed is the argument. The value is passed during the class instantiation.

     self.breed = breed
Now we have created two instances of the Dog class. With two breed types, we can then access these attributes like this:     

In [18]:
sam.breed

'Lab'

In [19]:
frank.breed

'Huskie'

Note how we don't have any parenthesis after breed, this is because it is an attribute and doesn't take any arguments.

In Python there are also class object attributes. These Class Object Attributes are the same for any instance of the class. For example, we could create the attribute species for the Dog class. Dogs (regardless of their breed,name, or other attributes will always be mammals. We apply this logic in the following manner:

In [129]:
class Dog(object):
    
    # Class Object Attribute
    species = 'mammal'
    
    def __init__(self, breed, name):
        self.breed = breed
        self.name = name

In [130]:
sam = Dog('Lab','Sam')

In [131]:
sam.name

'Sam'

Note that the Class Object Attribute is defined outside of any methods in the class. Also by convention, we place them first before the init.

In [133]:
sam.species

'mammal'

In [134]:
Dog.species

'mammal'

## Methods

Methods are functions defined inside the body of a class. They are used to perform operations with the attributes of our objects. Methods are essential in encapsulation concept of the OOP paradigm. This is essential in dividing responsibilities in programming, especially in large applications.

You can basically think of methods as functions acting on an Object that take the Object itself into account through its self argument.

Let's see the explanation and go through an example of creating a Circle class

<img src="../images/icon/Concept-Alert.png" alt="Concept-Alert" style="width: 100px;float:left; margin-right:15px"/>
<br />
### Explanation
***
- **`self`** is similar to the **`this`** pointer in other languages, except that (1) it needs to be explicitly passed as the first parameter of the instance method, and (2) it is not a reserved keyword

- The **`__init__`** method is an initializer (_not_ constructor) and called on instantiation

- The **`__str__`** method is equivalent to toString()

- The **`__repr__`** method defines how the object is represented on console

In [140]:
class Circle(object):
    '''
    This class if for Circle
    arg1: radius
    return ares of circle
    '''
    pi = 3.14

    # Circle get instantiated with a radius (default is 1)
    def __init__(self, radius=1):
        self.radius = radius 

    # Area method calculates the area. Note the use of self.
    def area(self):
        return self.radius * self.radius * Circle.pi

    # Method for resetting Radius
    def setRadius(self, radius):
        self.radius = radius

    # Method for getting radius (Same as just calling .radius)
    def getRadius(self):
        return self.radius


c = Circle()

c.setRadius(2)
print('Radius is: ',c.getRadius())
print('Area is  : ',c.area())

Radius is:  2
Area is  :  12.56


In [141]:
print(c.__str__)
print(c.__doc__)
print(c.__class__)
print(c.__init__)
print(c.__repr__)

<method-wrapper '__str__' of Circle object at 0x7fde30a7b438>

    This class if for Circle
    arg1: radius
    return ares of circle
    
<class '__main__.Circle'>
<bound method Circle.__init__ of <__main__.Circle object at 0x7fde30a7b438>>
<method-wrapper '__repr__' of Circle object at 0x7fde30a7b438>


<img src="../images/icon/ppt-icons.png" alt="ppt-icons" style="width: 100px;float:left; margin-right:15px"/>
<br />
Great! Notice how we used self. notation to reference attributes of the class within the method calls. Review how the code above works and try creating your own method

In [5]:
class Employee:
    'Common base class for all employees'
    empCount = 0

    def __init__(self, name, salary):
        self.name = name
        self.salary = salary
        Employee.empCount += 1
   
    def displayCount(self):
        print("Total Employee %d" % Employee.empCount)

    def displayEmployee(self):
        print("Name : ", self.name,  ", Salary: ", self.salary)

In [9]:
Employee.__doc__

'Common base class for all employees'

In [6]:
"This would create first object of Employee class"
emp1 = Employee("Zara", 2000)
"This would create second object of Employee class"
emp2 = Employee("Manni", 5000)

In [8]:
emp1.displayEmployee()
emp2.displayEmployee()
print("Total Employee %d" % Employee.empCount)

Name :  Zara , Salary:  2000
Name :  Manni , Salary:  5000
Total Employee 2


In [12]:
print("Employee.__doc__:", Employee.__doc__)
print("Employee.__name__:", Employee.__name__)
print("Employee.__module__:", Employee.__module__)
print("Employee.__bases__:", Employee.__bases__)
print("Employee.__dict__:", Employee.__dict__)

Employee.__doc__: Common base class for all employees
Employee.__name__: Employee
Employee.__module__: __main__
Employee.__bases__: (<class 'object'>,)
Employee.__dict__: {'empCount': 2, '__doc__': 'Common base class for all employees', '__dict__': <attribute '__dict__' of 'Employee' objects>, 'displayEmployee': <function Employee.displayEmployee at 0x7fcd1c2bec80>, '__weakref__': <attribute '__weakref__' of 'Employee' objects>, 'displayCount': <function Employee.displayCount at 0x7fcd1c2bed08>, '__init__': <function Employee.__init__ at 0x7fcd1c2beea0>, '__module__': '__main__'}


# Inheritance
Inheritance is a way to form new classes using classes that have already been defined. The newly formed classes are called derived classes, the classes that we derive from are called base classes. Important benefits of inheritance are code reuse and reduction of complexity of a program. The derived classes (descendants) override or extend the functionality of base classes (ancestors).


In [ ]:
# Syntax for derived class
class SubClassName (ParentClass1[, ParentClass2, ...]):
    'Optional class documentation string'
    class_suite

In [ ]:
### Generic inheritense example

In [14]:
class Parent:        # define parent class
    parentAttr = 100
    def __init__(self):
        print("Calling parent constructor")

    def parentMethod(self):
        print('Calling parent method')

    def setAttr(self, attr):
        Parent.parentAttr = attr

    def getAttr(self):
        print("Parent attribute :", Parent.parentAttr)

class Child(Parent): # define child class
    def __init__(self):
        print("Calling child constructor")

    def childMethod(self):
        print('Calling child method')

In [15]:
c = Child()          # instance of child
c.childMethod()      # child calls its method
c.parentMethod()     # calls parent's method
c.setAttr(200)       # again call parent's method
c.getAttr()          # again call parent's method

Calling child constructor
Calling child method
Calling parent method
Parent attribute : 200


In [17]:
print(Parent.parentAttr)
print(Child.parentAttr)

200
200


In [24]:
## Check the class relationship
print(issubclass(Child, Parent))
print(issubclass(Parent, Child))
print()

print(isinstance(c, Child))
print(isinstance(c, Parent))
p = Parent()
print(isinstance(p, Parent))
print(isinstance(p, Child))

True
False

True
True
Calling parent constructor
True
False


### Overriding Methods
- You can always override your parent class methods. One reason for overriding parent's methods is because you may want special or different functionality in your subclass.

In [27]:
class Parent:        # define parent class
    def myMethod(self):
        print('Calling parent method')

class Child(Parent): # define child class
    def myMethod(self):
        print('Calling child method')

c = Child()          # instance of child
c.myMethod()         # child calls overridden method

Calling child method


In [28]:
class Parent:        # define parent class
    def myMethod(self):
        print('Calling parent method')

class Child(Parent): # define child class
    pass

c = Child()          # instance of child
c.myMethod()         # child calls overridden method

Calling parent method


In [43]:
class Employee(object):
    def __init__(self, name, pay_rate):
        self.name = name
        self.pay_rate = pay_rate

    def displayEmployee(self):
        return "Name : ", self.name,  ", Salary: ", self.pay_rate

    def pay(self, hours_worked):
        return self.pay_rate * hours_worked

class Manager(Employee):
    def __init__(self, name, pay_rate, is_salaried):
        super(Manager, self).__init__(name, pay_rate)
        self.is_salaried = is_salaried

    def displayManager(self):
        #return Employee.__str(self) + " salaried: ",(self.is_salaried)
        return "Name : ", self.name,  ", Salary: ", self.pay_rate, " salaried: ",(self.is_salaried)

    # Override method
    def pay(self, hours_worked):
        if self.is_salaried:
            return self.pay_rate
        else:
            return super(Manager, self).pay(hours_worked)

In [38]:
e1 = Employee("John Jones", 10.00)
print(e1.name, ",", e1.pay_rate)
print(e1.displayEmployee())
print("Gross pay: ", e1.pay(40))

John Jones , 10.0
('Name : ', 'John Jones', ', Salary: ', 10.0)
Gross pay:  400.0


In [68]:
e1 = Employee("John Jones", 10.00)
print("Gross pay: ", e1.pay(40))

m1 = Manager("Jane Smith", 1200, True)
print(m1.displayEmployee())
print(m1.displayManager())
print("Gross pay: ", m1.pay(40))

m2 = Manager("Jim Brown", 20.00, False)
print("Gross pay: ", m2.pay(40))

Gross pay:  400.0
('Name : ', 'Jane Smith', ', Salary: ', 1200)
('Name : ', 'Jane Smith', ', Salary: ', 1200, ' salaried: ', True)
Gross pay:  1200
Gross pay:  800.0


### Encapsulation

In [69]:
class Computer:

    def __init__(self):
        self.__maxprice = 900

    def sell(self):
        print("Selling Price: {}".format(self.__maxprice))

    def setMaxPrice(self, price):
        self.__maxprice = price        

In [72]:
c = Computer()
c.sell()

# change the price
c.__maxprice = 1000
c.sell()

# using setter function
c.setMaxPrice(1000)
c.sell()

Selling Price: 900
Selling Price: 900
Selling Price: 1000


### Polymorphism

In [74]:
class Animal:
    def __init__(self, name):    # Constructor of the class
        self.name = name
    def talk(self):              # Abstract method, defined by convention only
        raise NotImplementedError("Subclass must implement abstract method")

class Cat(Animal):
    def talk(self):
        return 'Meow!'

class Dog(Animal):
    def talk(self):
        return 'Woof! Woof!'

animals = [Cat('Missy'),
           Cat('Mr. Mistoffelees'),
           Dog('Lassie')]

for animal in animals:
    print(animal)
    print(animal.name + ': ' + animal.talk())

Missy: Meow!
Mr. Mistoffelees: Meow!
Lassie: Woof! Woof!


## Duck Typing
***
- It is a feature of dynamic languages
- Central idea: If it walks like a duck and quacks like a duck then treat it like a duck.
- This is why Python doesn't have "interfaces", just "protocols"

In [76]:
class Duck:
    def quack(self):
        print("Quaaaaaack!")
    def feathers(self):
        print("The duck has white and gray feathers.")
    def name(self):
        print("ITS A DUCK NO NAME")

class Person:
    def quack(self):
        print("The person imitates a duck.")
    def feathers(self):
        print("The person takes a feather from the ground and shows it.")
    def name(self):
        print("John Smith")

def in_the_forest(duck):
    duck.quack()
    duck.feathers()
    duck.name()

def game():
    for element in [ Duck() , Person()]:
        in_the_forest(element)
        print()

game()

Quaaaaaack!
The duck has white and gray feathers.
ITS A DUCK NO NAME

The person imitates a duck.
The person takes a feather from the ground and shows it.
John Smith



- In polymorphism we see subclass (Cat and Dog) inheriting from the parent class (Animal) and overriding the method Talk.
- In case of duck typing we don’t create a subclass instead new class is created with method having same name but different function.

Great! you should have a basic understanding of how to create your own objects with class in Python. You will be utilizing this heavily in your upcomming projects!

<img src="../images/icon/Concept-Alert.png" alt="Concept-Alert" style="width: 100px;float:left; margin-right:15px"/>
<br />
## Instance vs Class vs Static Methods
***
- Instance methods have access to the **instance** of the class
- Class methods have access to the **class** (classes are also objects in Python), but not instances. This is similar to the static methods in Java/C#
- Static methods have no access to either instances or classes. They are more like plain functions, just bounded with the class for scoping

In [ ]:
class MyClass(object):
    
    def __init__(self, x =10):
        self.x = x
    
    def instance_method(self):
        print(self.x)
        print('instance method called', self)

    @classmethod
    def class_method(cls):
        #print(x)
        print('class method called', cls)

    @staticmethod
    def static_method():
        print('static method called')


https://realpython.com/instance-class-and-static-methods-demystified/

In [67]:
# Instance Method
obj = MyClass()

obj.instance_method()
MyClass.instance_method(obj)
obj.__class__

10
instance method called <__main__.MyClass object at 0x7fcd1c20a518>
10
instance method called <__main__.MyClass object at 0x7fcd1c20a518>


__main__.MyClass

In [47]:
obj.class_method()

class method called <class '__main__.MyClass'>


In [50]:
obj.static_method()

static method called


In [51]:
MyClass.class_method()

class method called <class '__main__.MyClass'>


In [52]:
MyClass.static_method()

static method called


### Talk about modules
- different ways of running a program
- passing arg while running a program
- import from other file creating a file
- comment in python - single line vs multiline

# Python Intermediate: NumPy Basics

### Why use Numpy

- https://www.youtube.com/watch?v=EEUXKG97YRw
- https://jakevdp.github.io/PythonDataScienceHandbook/

In [80]:
import numpy as np
import sys

sys.version
print(np.version.version)
# out: '1.6.2'

size = int(1E6)

%timeit for x in range(size): x ** 2
# out: 10 loops, best of 3: 136 ms per loop

%timeit for x in range(size): x ** 2
# out: 10 loops, best of 3: 88.9 ms per loop

# avoid this
%timeit for x in np.arange(size): x ** 2
#out: 1 loops, best of 3: 1.16 s per loop

# use this
%timeit np.arange(size) ** 2
#out: 100 loops, best of 3: 19.5 ms per loop

1.14.2
394 ms ± 3.56 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
375 ms ± 480 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
301 ms ± 1.73 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
10.2 ms ± 50.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


![caption](../images/numpy-logo.jpg)
***
NumPy, which stands for Numerical Python, is a library consisting of multidimensional array objects and a collection of routines for processing those arrays. Using NumPy, mathematical and logical operations on arrays can be performed. Let's along with the basics of NumPy such as its architecture and environment. It also discusses the various array functions, types of indexing, etc.

We can install numpy by `pip install numpy`

- Main object: **`ndarray`**

<img src="../images/icon/Technical-Stuff.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />
# ndarray
***

The most important object defined in NumPy is an N-dimensional array type called ndarray. It describes the collection of items of the same type. Items in the collection can be accessed using a zero-based index.

Every item in an ndarray takes the same size of block in the memory. Each element in ndarray is an object of data-type object (called dtype).Any item extracted from ndarray object (by slicing) is represented by a Python object of one of array scalar types. An instance of ndarray class can be constructed by different array creation routines described later.

You import the function in python by calling `import numpy`. The basic ndarray is created using an array function in NumPy as follows −

In [29]:
import numpy
numpy.array

<function numpy.core.multiarray.array>


<img src="../images/icon/Concept-Alert.png" alt="Concept-Alert" style="width: 100px;float:left; margin-right:15px"/>
<br />
## How do I create Arrays in Python?
***
* Create an array from a regular Python list or tuple using the array function. 

* The type of the resulting array is deduced from the type of the elements in the sequences

It creates an ndarray from any object exposing array interface, or from any method that returns an array.

In [30]:
numpy.array(object, dtype = None, copy = True, order = None, subok = False, ndmin = 0)

array(<type 'object'>, dtype=object)

In [ ]:
np.array

In [31]:
import numpy as np

# From list: 1d array
my_list = [10, 20, 30]
np.array(my_list)

array([10, 20, 30])

In [32]:
# From list: 2d array

list_of_lists =  [[5, 10, 15], [20, 25, 30], [35, 40, 45]]
np.array(list_of_lists)

array([[ 5, 10, 15],
       [20, 25, 30],
       [35, 40, 45]])

In [33]:
type(np.array(list_of_lists))

numpy.ndarray

An example of how does n-dimensional looks

## Types

![NumPy Array Types](../images/numpy-types1.jpg)

`ndarray` is also known by the alias `array`. Note that `numpy.array` is not the same as the Standard Python Library class `array.array`, which only handles one-dimensional arrays and offers less functionality. The more important attributes of an `ndarray` object are:

***ndarray.ndim***
the number of axes (dimensions) of the array. In the Python world, the number of dimensions is referred to as rank.

***ndarray.shape***
the dimensions of the array. This is a tuple of integers indicating the size of the array in each dimension. For a matrix with n rows and m columns, `shape` will be `(n,m)`. The length of the `shape` tuple is therefore the rank, or number of dimensions,`ndim`.

***ndarray.size***
the total number of elements of the array. This is equal to the product of the elements of shape.

***ndarray.dtype***
an object describing the type of the elements in the array. One can create or specify dtype’s using standard Python types. Additionally NumPy provides types of its own. numpy.int32, numpy.int16, and numpy.float64 are some examples.

***ndarray.reshape***
Returns an array containing the same data with a new shape.

<img src="../images/icon/Technical-Stuff.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />
## numpy.dtype
***
<br/>
The data type or dtype describes the kind of elements that are contained within the array.

* **bool**: Boolean values
<br/><br/>

* **int**: Integer values. Can be int16, int32, or int64.


* **float**: Floating point values. Can be float16, float32, or float64.
<br/><br/>


* ** string**: Text. Can be string or unicode (this distinction is greatly simplified in Python 3)

<img src="../images/icon/ppt-icons.png" alt="ppt-icons" style="width: 100px;float:left; margin-right:15px"/>
<br />
***
## Let's try it ourselves!
***
### Create a vector from the list [10, 20, 30]. Print the dtype and shape.

In [34]:
my_list = [10, 20, 30]

arr = np.array(my_list)

print(arr.dtype)
print(arr.shape)

int64
(3,)


In [81]:
my_list = [10, 20, 30.0]

arr = np.array(my_list)

print(arr.dtype)
print(arr.shape)

float64
(3,)


In [83]:
my_list = [10, 20, 30]

arr = np.array(my_list)

print(arr.dtype)
print(arr.shape)

<U21
(4,)


<img src="../images/icon/ppt-icons.png" alt="ppt-icons" style="width: 100px;float:left; margin-right:15px"/>
<br />
***
### Create a matrix from the list of lists [[5.3, 10.2, 15.1], [20.4, 25.3, 30.9], [35.4, 40.1, 45.6]]. Print the dtype and shape. 

### Important Concepts
***
#### Rank

NumPy’s main object is the homogeneous multidimensional array. It is a table of elements (usually numbers), all of the same type, indexed by a tuple of positive integers. In NumPy dimensions are called axes. The number of axes is rank.

For example, the coordinates of a point in 3D space [1, 2, 1] is an array of rank 1, because it has one axis. That axis has a length of 3. 

In the example below, the array has rank 2 (it is 2-dimensional). The first dimension (axis) has a length of 2, the second dimension has a length of 6.

In [84]:
a =  np.array([[1, 2, 3,4,5,6],[7,8,9,10,11,12]]) 
print(a)

[[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]]


In [36]:
print(a.shape)

(2, 6)


In [38]:
print(np.ndim(a))

2


In [86]:
a.size

12

# NumPy Built-in methods

## `arange`
***
arange(**[start,]** ***stop[, step,][, dtype]***) : Returns an array with evenly spaced elements as per the interval. The interval mentioned is half opened i.e. **[Start, Stop)** (similar to the Python **`range()`** function).

In [39]:
import numpy as np

np.arange(0, 10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

## `zeros` and `ones`
***
Generate arrays of all zeros and ones

In [40]:
np.zeros((2, 3))

array([[ 0.,  0.,  0.],
       [ 0.,  0.,  0.]])

In [87]:
np.zeros(5)

array([0., 0., 0., 0., 0.])

In [41]:
np.ones((2, 5))

array([[ 1.,  1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.,  1.]])

## `eye`
***
Creates an identity matrix of given size

In [42]:
np.eye(4)

array([[ 1.,  0.,  0.,  0.],
       [ 0.,  1.,  0.,  0.],
       [ 0.,  0.,  1.,  0.],
       [ 0.,  0.,  0.,  1.]])

## `linspace`
***
Linspace: Return **evenly spaced** numbers over a specified interval.

    linspace(start, stop, num=50, endpoint=True, retstep=False)

* Will return `num` number of values
* Equally spaced samples in the closed interval [start, stop] or the half-open interval [start, stop)
* Closed or half-open interval depends on whether 'endpoint' is True or False.

In [88]:
# divide into 7 interval from 0 to 10
np.linspace(0, 10, 7, endpoint=False)

array([0.        , 1.42857143, 2.85714286, 4.28571429, 5.71428571,
       7.14285714, 8.57142857])

<img src="../images/icon/Concept-Alert.png" alt="Concept-Alert" style="width: 100px;float:left; margin-right:15px"/>
<br />
## How do I generate Random Numbers?
***
<br/>
Numpy also has lots of ways to create random number arrays of given shape

- **`rand`**:

`numpy.random.rand(d0, d1, …, dn)` Create an array of the given shape and populate it with random samples from a **uniform distribution**

- **`randn`**: 

`numpy.random.randn(d0, d1, …, dn)`creates an array of specified shape and fills it with random values as per **standard normal distribution**.

If positive arguments are provided, randn generates an array of shape (d0, d1, …, dn), filled with random floats sampled from a univariate “normal” (Gaussian) distribution of mean 0 and variance 1 (if any of the d_i are floats, they are first converted to integers by truncation).

A single float randomly sampled from the distribution is returned if no argument is provided.

- **`randint`**: 

Return random integers from the “discrete uniform” distribution of the specified dtype in the “half-open” interval [low, high). If high is None (the default), then results are from [0, low).

In [44]:
# random number (uniform distribution) array of shape (5, 5)

np.random.rand(3, 4)

array([[ 0.98723719,  0.27977444,  0.36217167,  0.51239744],
       [ 0.88249281,  0.76137804,  0.39485289,  0.09073534],
       [ 0.43164014,  0.41684202,  0.22017851,  0.08392155]])

In [45]:
# random number (standard normal distribution) array of shape (2, 3)

print (np.random.randn(2, 3))

[[-0.50712953 -0.0459861   0.90274606]
 [ 0.54849607  0.17119001 -0.58995648]]


In [98]:
# 10 random integers between 4 (inclusive) to 40 (exclusive)

np.random.randint(4, 40, 10)

array([26,  7, 31, 39, 11, 23, 32, 17, 22, 19])

In [96]:
np.random.seed(40)
np.random.randint(4, 40, 10)

array([10, 31, 11,  5, 16, 11, 23, 35, 14, 23])

In [101]:
# 12 random integers upto 50 (exclusive). This makes the start value default to 0.
# The size parameter dictates the return array shape

np.random.randint(50, size=(3,4))

array([[46, 25, 19, 34],
       [ 6, 19, 21, 36],
       [37, 37, 44, 10]])

<img src="../images/icon/ppt-icons.png" alt="ppt-icons" style="width: 100px;float:left; margin-right:15px"/>
<br />
***
Lets try to create a ndimensions array from randomly genrated numbers of (3,4), then rehape the array to (4,3)


# Analyzing the Weather using NumPy

<center><img src="../images/weather.jpg" alt="Weather" style="width: 350px;"/></center>
Now it's time to use some them to learn data manipulation by analyzing a weather data set. As they say

We'll be working with **weather_small_2012.csv**, which contains weather data for each hour in 2012.
Since weather_small_2012.csv is a csv file, rows are separated by line breaks, and columns are
separated by commas:

```
Date/Time,Temp (C),Dew Point Temp (C),Rel Hum (%),Wind Spd (km/h),Visibility (km),Stn Press (kPa)
2012-01-01 00:00:00,-1.8,-3.9,86,4,8.0,101.24
2012-01-01 01:00:00,-1.8,-3.7,87,4,8.0,101.24
2012-01-01 02:00:00,-1.8,-3.4,89,7,4.0,101.26
2012-01-01 03:00:00,-1.5,-3.2,88,6,4.0,101.27
```

**To read csv file, we use:**

    numpy.genfromtxt(fileName, delimeter=",")

In [48]:
# read csv file
weather = np.genfromtxt("../data/weather_small_2012.csv", delimiter=",")

print (weather.dtype)
print (weather)

float64
[[    nan     nan     nan ...,     nan     nan     nan]
 [    nan   -1.8    -3.9  ...,    4.      8.    101.24]
 [    nan   -1.8    -3.7  ...,    4.      8.    101.24]
 ..., 
 [    nan   -0.5    -1.5  ...,   28.      4.8    99.95]
 [    nan   -0.2    -1.8  ...,   28.      9.7    99.91]
 [    nan    0.     -2.1  ...,   30.     11.3    99.89]]


Many items in this dataset are nan.

* The entire first row is nan – headers are String.
* Some of the numbers are written like 1.98600000e+03.

The data type of world_milk is float. Because all of the values in a NumPy array have to have the same
data type, NumPy attempted to convert all of the columns to floats when they were read in.

** Reading In The Data Properly **

***
To read world_milk.csv file properly we will have to use correct data type and skip the header.
* genfromtxt() default dtype is float, it converts non-numeric value to nan (not a number)
* To avoid nan, we read values as |S20 (String of length 20) 

In [49]:
weather = np.genfromtxt("../data/weather_small_2012.csv", dtype='|S20', skip_header=1, delimiter=",")

print (weather.dtype)
print (weather[0])

|S20
['2012-01-01 00:00:00' '-1.8' '-3.9' '86' '4' '8.0' '101.24']


In [50]:
# Create an array of temperatures from the data set

temperatures = weather[:,1].astype(np.float16)
print(temperatures)

dew_point_temperatures = weather[:,2].astype(np.float16)
print(dew_point_temperatures)

[-1.79980469 -1.79980469 -1.79980469 ..., -0.5        -0.19995117  0.        ]
[-3.90039062 -3.69921875 -3.40039062 ..., -1.5        -1.79980469
 -2.09960938]


<img src="../images/icon/Concept-Alert.png" alt="Concept-Alert" style="width: 100px;float:left; margin-right:15px"/>
<br />
# Operations with NumPy arrays
<br/>
***
NumPy provides a lot of built-in functionality for working with arrays.
**The important concepts to remember are**
- Any operation with a scalar number or a scalar function will cause that operation being computed for each element
- Any operation with two **compatible** (eg.: same shape) arrays will cause one-to-one element computations

<img src="../images/icon/Maths-Insight.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />
## Arithmetic (1/2)
***
### Vector Arithmetic
- All operations between arrays are **element-wise**
- This means that if you multiply two 2d vectors, it will **NOT** perform matrix multiplication

<img src="../images/icon/Maths-Insight.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />
## Arithmetic (2/2)
***
### Scalar Arithmetic
- Any operation of an array with a scalar will result in **element-wise** computation of that operation
- For example **`my_array + 2`** is the same as adding 2 to each element of array

### Calculate the Temperatures from the weather dataset in Farenheit

In [51]:
farenheit = (temperatures * 9 / 5) + 32
farenheit

array([ 28.765625,  28.765625,  28.765625, ...,  31.09375 ,  31.640625,
        32.      ], dtype=float16)

In [52]:
# Using default Python list
farenheit2 = [(celcius * 9 / 5) + 32 for celcius in temperatures]
farenheit2[:5]

[28.760351562499999,
 28.760351562499999,
 28.760351562499999,
 29.300000000000001,
 29.300000000000001]

### Addition

In [53]:
# Total temperature

# Vector Addition
print(temperatures + dew_point_temperatures)

# Scalar Addition
print(temperatures + 100)

[-5.69921875 -5.5        -5.19921875 ..., -2.         -2.         -2.09960938]
[  98.1875   98.1875   98.1875 ...,   99.5      99.8125  100.    ]


### Division

In [102]:
array1 = np.arange(1, 10, dtype=np.float16).reshape(3, 3)
array2 = np.arange(100, 109, dtype=np.float16).reshape(3, 3)
print()

print(array1)
print(array2)
print()

print(array2 / array1)  # Vector Division
print(array2 / 3)    # Scalar Division


[[1. 2. 3.]
 [4. 5. 6.]
 [7. 8. 9.]]
[[100. 101. 102.]
 [103. 104. 105.]
 [106. 107. 108.]]

[[100.     50.5    34.   ]
 [ 25.75   20.8    17.5  ]
 [ 15.14   13.375  12.   ]]
[[33.34 33.66 34.  ]
 [34.34 34.66 35.  ]
 [35.34 35.66 36.  ]]


## Comparison

Comparing two numpy arrays for equality, element-wise

In [55]:
# Find those temperatures that are above 0 degrees Celcius

greater_than_0 = temperatures > 0

print(temperatures)
print(greater_than_0)

print(type(greater_than_0))
print(greater_than_0.dtype)

[-1.79980469 -1.79980469 -1.79980469 ..., -0.5        -0.19995117  0.        ]
[False False False ..., False False False]
<type 'numpy.ndarray'>
bool


In [56]:
# multiple conditions
arr = np.array([[1,2,3],[4,5,6],[7,8,9]])

two_or_five = (arr == 2) | (arr == 5)
print(two_or_five)

[[False  True False]
 [False  True False]
 [False False False]]


In [57]:
arr1 = np.random.randint(1, 10, 6).reshape(2, 3)
arr2 = np.random.randint(1, 10, 6).reshape(2, 3)

print(arr1)
print(arr2)

print(arr1 >= arr2)

[[4 3 1]
 [3 7 6]]
[[8 2 9]
 [1 8 5]]
[[False  True False]
 [ True False  True]]


<img src="../images/icon/Technical-Stuff.png" alt="Concept-Alert" style="width: 100px;float:left; margin-right:15px"/>
<br />
## Aggregation 
* **`sum()`:** Computes the sum of all the elements in a vector, or the sum along a dimension in a matrix.
* **`mean()`:** Computes the average of all the elements in a vector, or the average along a dimension in a matrix.
* **`max()`/`min()`:** Identifies the maximum/minimum value among all the elements in a vector, or along a dimension in a matrix.
* **`argmax()`/`argmin()`:** Returns the index of maximum/minimum element.

In [103]:
# Find max, min, mean temperature
print('Max: ', temperatures.max())
print('Min: ', temperatures.min())
print('Mean: ', temperatures.mean())
print()

# Find index of max/min temperature
print('Argmax: ', temperatures.argmax())
print('Argmin: ', temperatures.argmin())

NameError: name 'temperatures' is not defined

<img src="../images/icon/Maths-Insight.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />
## Mathematical Functions
***
Standard mathematical functions like `sin`, `cos`, `ceil`, etc are available in NumPy in vectorized form.

In [108]:
arr = np.random.randint(100, size=9).reshape(3, 3)

print(arr)
print('\n')

print(np.sin(arr))

[[45  7 10]
 [61 19 68]
 [25 61 98]]


[[ 0.85090352  0.6569866  -0.54402111]
 [-0.96611777  0.14987721 -0.89792768]
 [-0.13235175 -0.96611777 -0.57338187]]


<img src="../images/icon/Maths-Insight.png" alt="Technical-Stuff" style="width: 100px;float:left; margin-right:15px"/>
<br />
## Custom Vectorized Functions
***

In [109]:
x = np.random.randn(10)
print (x)

def my_func(x):
    return 1 if x > 0 else 0

vectorized_func = np.vectorize(my_func)
vectorized_func(x)

[-1.10186254  0.57113756  1.06144781  0.20549787  0.19285262 -0.94004378
  1.35758736  0.99040063 -0.62157919  3.04195129]


array([0, 1, 1, 1, 1, 0, 1, 1, 0, 1])

In [110]:
aa = x > 0

In [112]:
print(aa)
print(x[aa])

[False  True  True  True  True False  True  True False  True]
[0.57113756 1.06144781 0.20549787 0.19285262 1.35758736 0.99040063
 3.04195129]


# Further Reading
***
- Python Official Documentation: https://docs.python.org/
- NumPy documentation: http://www.numpy.org/

<img src="../images/icon/Recap.png" alt="Recap" style="width: 100px;float:left; margin-right:15px"/>
<br />
# In-session Recap Time
***
- Modular Programming in Python
    - Functions
    - OOP: Classes, Inheritance

- NumPy
    - Creating Arrays
    - Built-in Methods
    - Data Manipulation
    - Operations: Reshaping, Arithmetic, Aggregation, etc.

# Thank You
***
### Coming up next...

- Numpy Advanced: Indexing and Selection
- Introduction to Pandas

For more queries - Reach out to academics@greyatom.com 